<a href="https://colab.research.google.com/github/TMQ5/my_projects/blob/main/stc_TV_T1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# STC Jawwy

In [ ]:
from google.colab import drive
drive.mount("/content/drive")

Mounted at /content/drive


In [ ]:
%cd "/content/drive/My Drive/Task1_STC_TV/"
%ls


/content/drive/My Drive/Task1_STC_TV
'Colab Notebooks'/  'stc_TV_Data _Set_T1.xlsb'  'stc TV_T1.ipynb'   Untitled0.ipynb


In [ ]:
# install libraries that are not installed by default
!pip install pyxlsb

In [ ]:
# Import the required libraries
import pandas as pd
import pyxlsb
import numpy as np


# Jawwy dataset
The dataset consists of meta details about the movies and tv shows as genre.
Also details about Users activities, spent duration and if watching in High definition or standard definition.
You have to analyse this dataset to find top insights, findings and to solve the four tasks assigned to you.

In [ ]:
# Read the XLSB file
dataframe = pd.read_excel("stc_TV_Data _Set_T1.xlsb",sheet_name="Final_Dataset")

# Copy the dataset 'dataframe'
df = dataframe.copy()

In [ ]:
# check the data shape
dataframe.shape

(1048575, 13)

There are `1,048,575` rows and `13` columns in the dataset.

In [ ]:
# Display the first 5 rows
dataframe.head()

Column1  date_  user_id_maped           program_name  duration_seconds  \
0        1  42882          26138             100 treets                40   
1        3  42876           7946                 Moana                 17   
2        4  42957           7418  The Mermaid Princess                  8   
3        5  42942          19307  The Mermaid Princess                 76   
4        7  42923          15860             Churchill                 87   

  program_class  season  episode                              program_desc  \
0         MOVIE       0        0                    Drama Movie100 Streets   
1         MOVIE       0        0                 Animation MovieMoana (HD)   
2         MOVIE       0        0  Animation MovieThe Mermaid Princess (HD)   
3         MOVIE       0        0  Animation MovieThe Mermaid Princess (HD)   
4         MOVIE       0        0             Biography MovieChurchill (HD)   

  program_genre  series_title  hd         original_name  
0         Drama             0   0            100 treets  
1     Animation             0   1                 Moana  
2     Animation             0   1  The Mermaid Princess  
3     Animation             0   1  The Mermaid Princess  
4     Biography             0   1             Churchill

# Data Preprocessing on the input data


In [ ]:
# Drop the "Column1" using .drop
dataframe = dataframe.drop(columns=["Column1"])

In [ ]:
# Trim spaces in movies names to avoid misspellings in input data
dataframe['program_name'] = dataframe['program_name'].str.strip()

In [ ]:
# Read date column as date data type
dataframe['date_'] = pd.to_datetime(dataframe['date_'], unit='D', origin='30/12/1899')

In [ ]:
# Read numeric columns as numeric data types
dataframe[['duration_seconds', 'season','episode','series_title','hd']] = dataframe[['duration_seconds', 'season','episode','series_title','hd']].apply(pd.to_numeric)


In [ ]:
# Read string columns as string data types
dataframe[['user_id_maped', 'program_name','program_class','program_desc','program_genre','original_name']] = dataframe[['user_id_maped', 'program_name','program_class','program_desc','program_genre','original_name']].astype(str)

In [ ]:
# Display the dataset after applying data types
dataframe.head()

date_ user_id_maped          program_name  duration_seconds  \
0 2017-05-27         26138            100 treets                40   
1 2017-05-21          7946                 Moana                17   
2 2017-08-10          7418  The Mermaid Princess                 8   
3 2017-07-26         19307  The Mermaid Princess                76   
4 2017-07-07         15860             Churchill                87   

  program_class  season  episode                              program_desc  \
0         MOVIE       0        0                    Drama Movie100 Streets   
1         MOVIE       0        0                 Animation MovieMoana (HD)   
2         MOVIE       0        0  Animation MovieThe Mermaid Princess (HD)   
3         MOVIE       0        0  Animation MovieThe Mermaid Princess (HD)   
4         MOVIE       0        0             Biography MovieChurchill (HD)   

  program_genre  series_title  hd         original_name  
0         Drama             0   0            100 treets  
1     Animation             0   1                 Moana  
2     Animation             0   1  The Mermaid Princess  
3     Animation             0   1  The Mermaid Princess  
4     Biography             0   1             Churchill

In [ ]:
# Check if any column has null value in the dataset
dataframe.isnull().any()

date_               False
user_id_maped       False
program_name        False
duration_seconds    False
program_class       False
season              False
episode             False
program_desc        False
program_genre       False
series_title        False
hd                  False
original_name       False
dtype: bool

There are NO missing values in the dataset.

In [ ]:
# Describe the numeric values in the dataset
dataframe.describe()

date_  duration_seconds        season  \
count                        1048575      1.048575e+06  1.048575e+06   
mean   2017-10-04 00:23:20.346183936      1.230957e+03  1.342139e+00   
min              2017-03-14 00:00:00      2.000000e+00  0.000000e+00   
25%              2017-06-10 00:00:00      5.200000e+01  0.000000e+00   
50%              2017-10-14 00:00:00      1.190000e+02  1.000000e+00   
75%              2018-01-21 00:00:00      1.328000e+03  1.000000e+00   
max              2018-04-30 00:00:00      1.461329e+06  2.300000e+01   
std                              NaN      6.821058e+03  2.104095e+00   

            episode  series_title            hd  
count  1.048575e+06  1.048575e+06  1.048575e+06  
mean   6.157952e+00  1.205922e-02  3.862728e-01  
min    0.000000e+00  0.000000e+00  0.000000e+00  
25%    0.000000e+00  0.000000e+00  0.000000e+00  
50%    1.000000e+00  0.000000e+00  0.000000e+00  
75%    9.000000e+00  0.000000e+00  1.000000e+00  
max    2.820000e+02  1.000000e+00  1.000000e+00  
std    1.222015e+01  1.091504e-01  4.868946e-01

The statistical descriptive for numerical values:

`duration_seconds`:

- The maximum duration in seconds: **1,461,329**   

- The minimum duration in seconds: **2**  

- The mean duration in seconds: **around 1231**

- Standard deviation: **around 6821**


`season`:

- The maximum season number: **23 seasons**   

- The minimum season number: **0**  

- The mean seasons number: **around 1 "One" season**

- Standard deviation: **around 2 seasons**

`episode`:

- The maximum  episode number: **282**   

- The minimum  episode number: **0**  

- The mean  episodes number: **around 6 episodes**

- Standard deviation: **around 12 episodes**

`series_title	`:

- The maximum: **1**  *(Because it's a Binary Encoding)*

- The minimum: **0**  *(Because it's a Binary Encoding)*

- The mean: **around 0.012**

*(The proportion of values that represent "1" (has a series title) is about 1.21% of the total dataset. The rest of the data (Around 98.79%) represent the value "0"(has no series title)).*

- Standard deviation: **0.1091504**

*(The standard deviation value here indicates that most of the dataset is "0"(has no series title), and a small percentage of the dataset contains the value "1"(has a series title)).*

`hd`:

- The maximum: **1**   *(Because it's a Binary Encoding)*

- The minimum: **0**   *(Because it's a Binary Encoding)*

- The mean: **around 0.38627** *(38.63% is HD and the rest 61.37% is SD)*

- Standard deviation: **0.4868946**

# Task 1
##### You are required to work on task one to study and HD flag for available dataset

In [ ]:
# Make a copy of the dataframe for working on task 1
df=dataframe.copy()

In [ ]:
# Here we try to get the most watched movies (Total Views / Total Users Views / Total watch time)
# For series we concatenated the Session episode to differentiate between episodes

grouped=df.copy()
grouped.loc[grouped['program_class'] == 'SERIES/EPISODES', 'program_name'] = grouped['program_name']+'_SE'+grouped['season'].astype(str)+'_EP'+grouped['episode'].astype(str)
grouped = grouped.groupby(['program_name','program_class'])\
.agg({'user_id_maped': [('co1', 'nunique'),('co2', 'count')],\
      'duration_seconds': [('co3', 'sum')] }).reset_index()
grouped.columns = ['program_name','program_class','No of Users who Watched', 'No of watches', 'Total watch time in seconds']
grouped['Total watch time in houres']=grouped['Total watch time in seconds']/3600
grouped = grouped.drop(columns=['Total watch time in seconds'])
grouped = grouped.sort_values(by=['Total watch time in houres', 'No of watches','No of Users who Watched'], ascending=False).reset_index(drop=True)


In [ ]:
# Show the result
grouped.head(30)

program_name program_class  No of Users who Watched  \
0                       The Boss Baby         MOVIE                     3389   
1               The Amazing pider-Man         MOVIE                     1011   
2                     The Expendables         MOVIE                      853   
3                               Moana         MOVIE                     2173   
4                              Trolls         MOVIE                     2613   
5                                Bean         MOVIE                      949   
6                           The murfs         MOVIE                      867   
7                  Hotel Transylvania         MOVIE                      491   
8   Cloudy With a Chance of Meatballs         MOVIE                      683   
9         The Man With The Iron Fists         MOVIE                      707   
10                               Salt         MOVIE                      563   
11                           Unbroken         MOVIE                      625   
12                         ParaNorman         MOVIE                      614   
13                 Youm Maloosh Lazma         MOVIE                     1131   
14                          Ferdinand         MOVIE                     1278   
15                       White Chicks         MOVIE                      307   
16                      Jurassic Park         MOVIE                      504   
17                   The November Man         MOVIE                      494   
18                       Total Recall         MOVIE                      587   
19                         Robin Hood         MOVIE                      588   
20                     Public Enemies         MOVIE                      368   
21                     Daddy Day Camp         MOVIE                      263   
22                           Oblivion         MOVIE                      790   
23                              Blitz         MOVIE                      562   
24     War for the Planet of the Apes         MOVIE                      879   
25                         Inside Man         MOVIE                      532   
26                           Bad Boys         MOVIE                      438   
27                             Easy A         MOVIE                      513   
28                         Battleship         MOVIE                      634   
29                           Baywatch         MOVIE                     2062   

    No of watches  Total watch time in houres  
0           24047                 2961.350833  
1            2877                 1966.119167  
2            2119                 1961.159444  
3            8081                 1706.176944  
4           13793                 1601.023056  
5            3617                 1423.955000  
6            3132                 1342.141111  
7            1947                 1096.533611  
8            2076                  948.674722  
9            2505                  859.626389  
10           1082                  767.392778  
11           1429                  763.078333  
12           1746                  747.065556  
13           2278                  718.109722  
14           6817                  714.223056  
15            916                  711.840833  
16           1192                  693.394444  
17           1219                  679.492222  
18           1108                  661.820000  
19           1209                  643.935000  
20            716                  634.035000  
21            647                  625.338333  
22           1678                  609.391111  
23           1200                  570.521944  
24           2028                  567.597778  
25           1567                  560.386111  
26            871                  559.277500  
27            990                  557.068611  
28           1324                  552.857222  
29           7436                  548.995556

# Top10 Programs:

In [ ]:
# We import Visualization libraries
# You can ignore and use any other graphing libraries
import matplotlib.pyplot as plt # a comprehensive library for creating static, animated, and interactive visualizations
import plotly #a graphing library makes interactive, publication-quality graphs. Examples of how to make line plots, scatter plots, area charts, bar charts, error bars, box plots, histograms, heatmaps, subplots, multiple-axes, polar charts, and bubble charts.
import plotly.express as px
import plotly.graph_objects as go
from plotly.subplots import make_subplots

In [ ]:
# Plot top 10 Programs in total watch time in houres
fig = px.pie(grouped.head(10), values='Total watch time in houres', names='program_name',\
             hover_data=['program_class'],title='Top 10 programs in total watch time in houres')
fig.show()

The top 10 programs in total watch time in hours are:

1- `The Boss Baby` **Movie:** *By 18.7%, approximately 3,000 hours.*

2- `The Amazing Spider-Man` **Movie:** *By 12.4%, or 1,966 hours.*

3- `The Expendables` **Movie:** *By 12.4%, or 1,961 hours.*

4- `Moana` **Movie:** *By 10.8%, approximately 1,700 hours.*

5- `Trolls` **Movie:** *By 10.1%, approximately 1,600 hours.*

6- `Bean` **Movie:** *By 8.97%, approximately 1,400 hours.*

7- `The Murfs` **Movie:** *By 8.46%, approximately 1,300 hours.*

8- `Hotel Transylvania` **Movie:** *By 6.91%, approximately 1,000 hours.*

9- `Cloudy with a Chance of Meatballs` **Movie:** *By 5.98%, approximately 949 hours.*

10- `The Man with the Iron Fists` **Movie:** *By 5.42%, approximately 860 hours.*



In [ ]:
# Plot top 10 Programs in total number of watches
fig = px.pie(grouped.head(10), values='No of watches', names='program_name',\
             hover_data=['program_class'],title='Top 10 programs in Number of Watches')
fig.show()

The top 10 programs in number of watches are:

1- `The Boss Baby` **Movie:** *By 37.5%, approximately 24,000 watches.*

2- `Trolls` **Movie:** *By 21.5%, approximately 14,000 watches.*

3- `Moana` **Movie:** *By 12.6%, approximately 8,000 watches.*

4- `Bean` **Movie:** *By 5.63%, approximately 4,000 watches.*

5- `The Murfs` **Movie:** *By 4.88%, approximately 3,100 watches.*

6- `The Amazing Spider-Man` **Movie:** *By 4.48%, approximately 2,900 watches.*

7- `The Man with the Iron Fists` **Movie:** *By 3.9%, approximately 2,500 watches.*

8- `The Expendables` **Movie:** *By 3.3%, approximately 2,100 watches.*

9- `Cloudy with a Chance of Meatballs` **Movie:** *By 3.23%, or 2,076 watches.*

10- `Hotel Transylvania` **Movie:** *By 3.03%, approximately 2,000 watches.*



In [ ]:
# Plot top 10 Programs in total number of users who watched the programs
fig = px.pie(grouped.head(10), values='No of Users who Watched', names='program_name',\
             hover_data=['program_class'],title='Top 10 programs in Number of Users who Watched')
fig.show()

The top 10 programs in number of users who watched are:

1- `The Boss Baby` **Movie:** *By 24.7%, approximately 3,400 users.*

2- `Trolls` **Movie:** *By 19%, approximately 2,600 users.*

3- `Moana` **Movie:** *By 15.8%, approximately 2,200 users.*

4- `The Amazing Spider-Man` **Movie:** *By 7.36%, approximately 1,000 users.*

5- `Bean` **Movie:** *By 6.91%, approximately 950 users.*

6- `The Murfs` **Movie:** *By 6.31, approximately 870 users.*

7- `The Expendables` **Movie:** *By 6.21%, approximately 850 users.*

8- `The Man with the Iron Fists` **Movie:** *By 5.15%, or 707 users.*

9- `Cloudy with a Chance of Meatballs` **Movie:** *By 4.97%, or 680 users.*

10- `Hotel Transylvania` **Movie:** *By 3.57%, approximately 490 users.*



The previous top 10 programs were overall.

We will now find the top 10 programs based on whether is a movie or a series.

**FIRST: Moives**

In [ ]:
# Filtering and sorting the 'grouped' DataFrame to select the top 10 movies based on total watch time in hours,
# number of watches, and number of users who watched, in descending order of importance.
movie_class = grouped[grouped['program_class'] == 'MOVIE'].sort_values(by=['Total watch time in houres', 'No of watches', 'No of Users who Watched'], ascending=False).head(10)
movie_class

program_name program_class  No of Users who Watched  \
0                      The Boss Baby         MOVIE                     3389   
1              The Amazing pider-Man         MOVIE                     1011   
2                    The Expendables         MOVIE                      853   
3                              Moana         MOVIE                     2173   
4                             Trolls         MOVIE                     2613   
5                               Bean         MOVIE                      949   
6                          The murfs         MOVIE                      867   
7                 Hotel Transylvania         MOVIE                      491   
8  Cloudy With a Chance of Meatballs         MOVIE                      683   
9        The Man With The Iron Fists         MOVIE                      707   

   No of watches  Total watch time in houres  
0          24047                 2961.350833  
1           2877                 1966.119167  
2           2119                 1961.159444  
3           8081                 1706.176944  
4          13793                 1601.023056  
5           3617                 1423.955000  
6           3132                 1342.141111  
7           1947                 1096.533611  
8           2076                  948.674722  
9           2505                  859.626389

In [ ]:
# Plot top 10 movies in total watch time in houres
fig = px.bar(movie_class, x='Total watch time in houres', y='program_name', title="Total Watch Time in Houres")
fig.show()

In [ ]:
# Plot top 10 movies in total number of watches
fig = px.bar(movie_class, x='No of watches', y='program_name', title="No. of Watches")
fig.show()

In [ ]:
# Plot top 10 movies in total number of users who watched
fig = px.bar(movie_class, x='No of Users who Watched', y='program_name', title="No. of Users who Watched")
fig.show()

The movie class metrics (duration in hours, number of watches, and number of user who watches) are **EXACTLY** the same as we founded before in overall top 10 programs.

**SECOND: Series**

In [ ]:
# Filtering and sorting the 'grouped' DataFrame to select the top 10 series based on total watch time in hours,
# number of watches, and number of users who watched, in descending order.
series_class = grouped[grouped['program_class'] == 'SERIES/EPISODES'].sort_values(by=['Total watch time in houres', 'No of watches', 'No of Users who Watched'], ascending=False).head(10)

# Resetting the index for the filtered DataFrame to make it cleaner and more presentable.
series_class.reset_index(drop=True)

program_name    program_class  \
0           Blindspot     Erase Weary Youth_SE1_EP13  SERIES/EPISODES   
1  PAW Patrol  Pups Make A plash / Pups Fall Fest...  SERIES/EPISODES   
2        PAW Patrol  Pups ave The ea Turtles_SE1_EP2  SERIES/EPISODES   
3   PAW Patrol  Pups And The Kitty-Tastrophe_SE1_EP3  SERIES/EPISODES   
4  PAW Patrol  Pup Pup Goose / Pup Pup And Away_S...  SERIES/EPISODES   
5  Friends      The One With The ecret Closet_SE8...  SERIES/EPISODES   
6  PAW Patrol  Pup Pup Boogie / Pups In The Fog_S...  SERIES/EPISODES   
7  PAW Patrol  Pups On Ice / Pups And The now Mon...  SERIES/EPISODES   
8  PAW Patrol  Pups ave The Circus / Pup A Doodle...  SERIES/EPISODES   
9   Shimmer And hine  The First Wish (Part )_SE1_EP1  SERIES/EPISODES   

   No of Users who Watched  No of watches  Total watch time in houres  
0                      125            197                  516.015556  
1                      407           1774                  485.739722  
2                      334           1439                  478.470278  
3                      294           1296                  455.909722  
4                      268           1222                  448.070000  
5                       83            133                  445.568056  
6                      283           1334                  438.068889  
7                      256           1198                  423.900000  
8                      239           1099                  416.540556  
9                      427           1479                  413.285000

In [ ]:
# Plot top 10 series in total watch time in houres
fig = make_subplots(rows=1, cols=2, specs=[[{'type': 'bar'}, {'type': 'pie'}]])

fig.add_trace(go.Bar(x=series_class['program_name'], y=series_class['Total watch time in houres']),row=1, col=1)
fig.add_trace(go.Pie(labels=series_class['program_name'], values=series_class['Total watch time in houres']),row=1, col=2)

fig.update_layout(title_text='Top10 Series Based on Total Watch Time in Hours')

fig.show()

The top10 series in total watch time in hours are:

1- `Blindspot Erase Weary Youth SE1 EP13`: *By 11.4%, approximately 517 hours.*

2- `PAW Patrol Pups Make a Splash / Pups Fall Festival SE1 EP1`: *By 10.7%, approximately 486 hours.*

3- `PAW Patrol Pups ave The ea Turtles SE1 EP2`: *By 10.6%, approximately 480 hours.*

4- `PAW Patrol Pups And The Kitty-Tastrophe SE1 EP3`: *By 10.1%, approximately 460 hours.*

5- `PAW Patrol Pup Pup Goose / Pup Pup And Away SE1 EP5`: *By 9.91%, or 448 hours.*

6- `Friends The One With The ecret Closet SE8 EP14`: *By 9.85%, or 446 hours.*

7- `PAW Patrol Pup Pup Boogie / Pups In The Fog SE1 EP4`: *By 9.69%, approximately 438 hours.*

8- `PAW Patrol Pups On Ice / Pups And The now Monster SE1 EP6`: *By 9.38%, approximately 424 hours.*

9- `PAW Patrol Pups ave The Circus / Pup A Doodle Do SE1 EP7`: *By 9.21%, approximately 417 hours.*

10- `Shimmer And hine The First Wish (Part ) SE1 EP1`: *By 9.14%, approximately 413 hours.*



In [ ]:
# Plot top 10 series in total number of watches
fig = make_subplots(rows=1, cols=2, specs=[[{'type': 'bar'}, {'type': 'pie'}]])
sorted_series_class = series_class.sort_values(by='No of watches', ascending=False)

fig.add_trace(go.Bar(x=sorted_series_class['program_name'], y=sorted_series_class['No of watches']),row=1, col=1)
fig.add_trace(go.Pie(labels=sorted_series_class['program_name'], values=sorted_series_class['No of watches']),row=1, col=2)

fig.update_layout(title_text='Top10 Series Based on No. of Watches')

fig.show()

The top10 series in number of watches are:

1- `PAW Patrol Pups Make a Splash / Pups Fall Festival SE1 EP1`: *By 15.9%, approximately 1,700 watches.*

2- `Shimmer And hine The First Wish (Part ) SE1 EP1`: *By 13.2%, approximately 1,500 watches.*

3- `PAW Patrol Pups ave The ea Turtles SE1 EP2`: *By 12.9%, approximately 1,400 watches.*

4- `PAW Patrol Pup Pup Boogie / Pups In The Fog SE1 EP4`: *By 11.9%, approximately 1,330 watches.*

5- `PAW Patrol Pups And The Kitty-Tastrophe SE1 EP3`: *By 11.6%, approximately 1,300 watches.*

6- `PAW Patrol Pup Pup Goose / Pup Pup And Away SE1 EP5`: *By 10.9%, approximately 1,220 watches.*

7- `PAW Patrol Pups On Ice / Pups And The now Monster SE1 EP6`: *By 10.7%, approximately 1,200 watches.*

8- `PAW Patrol Pups ave The Circus / Pup A Doodle Do SE1 EP7`: *By 9.84%, approximately 1,100 watches.*

9- `Blindspot Erase Weary Youth SE1 EP13`: *By 1.76%, approximately 200 watches.*

10- `Friends The One With The ecret Closet SE8 EP14`: *By 1.19%, approximately 130 watches.*



In [ ]:
# Plot top 10 series in total number of users who watched
fig = make_subplots(rows=1, cols=2, specs=[[{'type': 'bar'}, {'type': 'pie'}]])
sorted_series_class = series_class.sort_values(by='No of Users who Watched', ascending=False)

fig.add_trace(go.Bar(x=sorted_series_class['program_name'], y=sorted_series_class['No of Users who Watched']),row=1, col=1)
fig.add_trace(go.Pie(labels=sorted_series_class['program_name'], values=sorted_series_class['No of Users who Watched']),row=1, col=2)

fig.update_layout(title_text='Top10 Series Based on No. of Users who Watched')

fig.show()

The top10 series in number of users who watched are:

1- `Shimmer And hine The First Wish (Part ) SE1 EP1`: *By 15.7%, approximately 430 users.*

2- `PAW Patrol Pups Make a Splash / Pups Fall Festival SE1 EP1`: *By 15%, approximately 410 users.*

3- `PAW Patrol Pups ave The ea Turtles SE1 EP2`: *By 12.3%, approximately 1,330 users.*

4- `PAW Patrol Pups And The Kitty-Tastrophe SE1 EP3`: *By 10.8%, approximately 300 users.*

5- `PAW Patrol Pup Pup Boogie / Pups In The Fog SE1 EP4`: *By 10.4%, approximately 280 users.*

6- `PAW Patrol Pup Pup Goose / Pup Pup And Away SE1 EP5`: *By 9.87%, approximately 270 users.*

7- `PAW Patrol Pups On Ice / Pups And The now Monster SE1 EP6`: *By 9.43%, approximately 260 users.*

8- `PAW Patrol Pups ave The Circus / Pup A Doodle Do SE1 EP7`: *By 8.8%, approximately 240 users.*

9- `Blindspot Erase Weary Youth SE1 EP13`: *By 4.6%, approximately 130 users.*

10- `Friends The One With The ecret Closet SE8 EP14`: *By 3.06%, approximately 80 users.*



# Customer Experience Study:

In [ ]:
# Here we try to study the customer experience against Program class
grouped=df.copy()
grouped = grouped.groupby('program_class')\
.agg({'user_id_maped': [('co1', 'nunique'),('co2', 'count')],\
      'duration_seconds': [('co3', 'sum')] }).reset_index()
grouped.columns = ['program_class','No of Users who Watched', 'No of watches', 'Total watch time in seconds']
grouped['Total watch time in houres']=grouped['Total watch time in seconds']/3600
grouped = grouped.drop(columns=['Total watch time in seconds'])
grouped = grouped.sort_values(by=['Total watch time in houres', 'No of watches','No of Users who Watched'], ascending=False).reset_index(drop=True)


In [ ]:
# Show the result
grouped.head()

program_class  No of Users who Watched  No of watches  \
0  SERIES/EPISODES                     3901         560174   
1            MOVIE                    11355         488401   

   Total watch time in houres  
0               255097.787500  
1               103444.145556

In [ ]:
# Plot the total watch time against total number of users and report your findings
# Creating a subplot layout with two pie charts side by side within the same figure.
fig = make_subplots(rows=1, cols=2, specs=[[{'type':'pie'}, {'type':'pie'}]])

# Adding the first pie chart which displays the total watch time in hours for each program class (movies and series).
fig.add_trace(go.Pie(labels=grouped['program_class'], values=grouped['Total watch time in houres'], name='Total Watch Time'),1, 1)

# Adding the second pie chart which displays the number of users who watched each program class (movies and series)
fig.add_trace(go.Pie(labels=grouped['program_class'], values=grouped['No of Users who Watched'], name='Number of Users'),1, 2)

# Updating the layout with a title and displaying the figure to show the distribution of watch time and user by program type.
fig.update_layout(title='Program Analysis by Type')
fig.show()

In [ ]:
# Creating a pie chart which displays the distribution of the number of watches by program class.
fig = px.pie(grouped, values='No of watches', names='program_class',\
             hover_data=['program_class'],title='No. of watches by program_class')

# Disabling the default sorting of pie chart segments to maintain the original order from the data.
fig.update_traces(sort=False)
fig.show()

In [ ]:
# Reshape the data for better visualization
melted_grouped = grouped.melt(id_vars= 'program_class', var_name= 'Metric', value_name= 'Value')

In [ ]:
# Creating a grouped bar chart to compare different metrics across program classes.
fig = px.bar(melted_grouped, x='program_class', y='Value', color='Metric', barmode='group', title='Comparison of program classes')
fig.show()

**Findings:**



From previous charts we can noticed that series class has a higher number of watches and a longer total watching time compared to movies.

> This means that series have many prolonged watching sessions, which may indicate continued demand and great interest in watching the episodes.

While the Movies class has more users who watched it, but less overall time than series.

> This may indicate that movies attract a wider audience but for one-time or short viewing periods, which reflects the nature of movies which are usually one complete watching rather than over a series of episodes like soap operas.



*It would be useful to do a deeper analysis to understand why users prefer movies despite spending less time watching them compared to series. This may include consideration of factors such as age, gender, geographic location, and cultural preferences.*





# Studying the Relationship of User Watches (watch resolution HD or SD):

In [ ]:
"""
TODO study the relation and user's behaviour Against HD flag, feel free to include any comparison
"""
resolution = df.copy()
resolution.loc[resolution['program_class'] == 'SERIES/EPISODES', 'program_name'] = resolution['program_name']+'_SE'+resolution['season'].astype(str)+'_EP'+resolution['episode'].astype(str)
resolution_grouped = resolution.groupby(['program_class', 'program_name'])\
  .agg({'user_id_maped': 'nunique',
        'hd': ['count', lambda x: (x==1).sum(), lambda x: (x==0).sum()],
        'duration_seconds': 'sum'}).reset_index()

resolution_grouped.columns = ['_'.join(col).strip() if col[1] else col[0] for col in resolution_grouped.columns.values]

resolution_grouped['Total watch time in hours'] = resolution_grouped['duration_seconds_sum'] / 3600
resolution_grouped.drop(columns=['duration_seconds_sum'], inplace=True)

columns = ['program_class', 'program_name', 'user_id_maped_nunique', 'hd_count', 'hd_<lambda_0>', 'hd_<lambda_1>', 'Total watch time in hours']
resolution_final = resolution_grouped[columns]

resolution_final.columns = ['Program Class', 'Program Name', 'No of Users who Watched', 'Total No of Watches', 'HD Watches', 'SD Watches', 'Total Watch Time in Hours']


In [ ]:
# Show the result
resolution_final.head(10)

Program Class           Program Name  No of Users who Watched  \
0         MOVIE          #FollowFriday                      316   
1         MOVIE  10 Days in a Madhouse                      301   
2         MOVIE             100 treets                      704   
3         MOVIE         101 Dalmatians                       85   
4         MOVIE         102 Dalmatians                      130   
5         MOVIE                   11.6                      212   
6         MOVIE              12 Rounds                        8   
7         MOVIE   12 Rounds : Reloaded                       32   
8         MOVIE              127 Hours                       40   
9         MOVIE               17 Again                       65   

   Total No of Watches  HD Watches  SD Watches  Total Watch Time in Hours  
0                  510         435          75                  53.269167  
1                  553           0         553                  99.112778  
2                 1851        1805          46                  88.611111  
3                  118           0         118                   8.333889  
4                  172           0         172                   8.954167  
5                  379           0         379                  92.149722  
6                   11           0          11                   0.173889  
7                   48           0          48                   9.039722  
8                   58           0          58                  14.331389  
9                   99           0          99                  17.302222

In [ ]:
# Plot a pie chart to visualize the distribution of watch quality based on the total number of watches for HD and SD.
values = resolution_final[['HD Watches', 'SD Watches']].sum().tolist()
names = ['HD', 'SD']

fig = px.pie(values=values, names=names, title='Distribution of Watch Quality')
fig.show()

The previous chart indicates that the number of HD watches quality is 405,036, representing 38.6% of the total watches, while the number of SD quality watches is 643,539, representing 61.4% of the total watches.

This precentage shows a preference for SD quality over HD, as there is a 58.88% difference between the number of SD watches and HD wathces, equivalent to 238,503 watches.

This significant difference suggests a user preference for this quality or cause to possibly other factors that prevent users from utilizing HD quality, such as internet connection quality constraints, the type of devices used that may not support HD quality, or data costs.

Based on this information, it might be advisable for STC to improve the user experience for the STC TV application and provide content that enhances the watching experience in SD quality.

In [ ]:
# Creating a heatmap to displays the correlation matrix of numerical data in 'resolution_final' DataFrame.
# Selecting only numeric data types from the DataFrame for correlation analysis.
numeric_df = resolution_final.select_dtypes(include=[np.number])

# Calculating the correlation matrix for the numeric columns in the DataFrame.
correlation_matrix = numeric_df.corr()


fig = px.imshow(correlation_matrix,
                labels=dict(x="Variables", y="Variables", color="Correlation Coefficient"),
                x=correlation_matrix.columns,
                y=correlation_matrix.columns,
                text_auto=True,
                title="Heatmap of Correlation Matrix")

fig.update_xaxes(side="bottom")

fig.show()

We can assert that there is a strong positive relationship between the number of users who watch programs on the STC TV app and the number of watches in HD quality; that is, as the number of users within the app who choose HD quality increases, so does the number of watches in this quality. Similarly, there is also a strong positive relationship in terms of the number of watches: as the number of watches for a particular program on the STC TV app increases, so does the number of watches that occur in HD quality (this indicates that the more popular programs tend to attract more watches in high quality).

As for the number of hours that app users spend watching programs, and the number of watches in both HD and SD qualities, there is a moderately positive relationship between them, although the number of watches in HD quality is slightly higher. This suggests that although watches in HD quality are not the only ones that affect the duration of watching, they have a noticeable impact.

In [ ]:
# Aggregating the total number of HD and SD watches by program class from the 'resolution_final' DataFrame.
grouped = resolution_final.groupby('Program Class')\
    .agg({'HD Watches': 'sum',
          'SD Watches': 'sum'}).reset_index()

# Creating a stacked bar chart to display the sums of HD and SD watches by program class.
fig = px.bar(grouped, x='Program Class', y=['HD Watches', 'SD Watches'],
             title= 'Watch Quality by Program Class',
             labels= {'value': 'Number of Watch', 'variable': 'Qulaity Type'})

fig.show()

The chart above shows that movies in HD quality attract more watches than series, while series in SD quality draw more wathces than movies. The high production quality of movies and the complex visual details they contain may enhance viewers' preference to watch them in HD, leading to an increase in the number of watches in this category with higher quality.

On the other hand, series are characterized by their longer duration and frequent watching, making them more suitable for watching in SD quality, especially in conditions of limited data or internet connectivity.